@copilot Remedy the comments for chatsnack/runtime/chat_completions_adapter.py above, and then also remedy the issues I am seeing running tests now:

chatsnack\tests\mixins\test_query.py::test_cleaned_chat_completion_model_fallback failed: chat = Chat(name='_ChatPrompt-2026-03-11_16-06-04-80a1c694-bce6-445f-b9f9-27f3b2856760', params=None, messages=[{'system': 'system'}])

    @pytest.mark.asyncio
    async def test_cleaned_chat_completion_model_fallback(chat):
        fake_completions = _FakeAsyncCompletions()
        chat.ai.aclient = SimpleNamespace(
            chat=SimpleNamespace(completions=fake_completions)
        )
        chat.system("system")
>       await chat._cleaned_chat_completion(chat.json)

tests\mixins\test_query.py:251: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
chatsnack\chat\mixin_query.py:340: in _cleaned_chat_completion
    normalized = await adapter.create_completion_a(messages=messages, **kwargs)
chatsnack\runtime\chat_completions_adapter.py:66: in create_completion_a
    return self._normalize_completion(response)
chatsnack\runtime\chat_completions_adapter.py:50: in _normalize_completion
    response_dict = self._to_dict(response)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

obj = <test_query._FakeResponse object at 0x00000257B1A79C60>

    @staticmethod
    def _to_dict(obj: Any) -> Dict[str, Any]:
        if obj is None:
            return {}
        if isinstance(obj, dict):
            return obj
        if hasattr(obj, "model_dump"):
            return obj.model_dump()
>       return dict(obj)
E       TypeError: '_FakeResponse' object is not iterable

chatsnack\runtime\chat_completions_adapter.py:26: TypeError
----
Same with test_cleaned_chat_completion_returns_message_object_for_tool_calls

--- a/chatsnack/runtime/chat_completions_adapter.py
+++ b/chatsnack/runtime/chat_completions_adapter.py
@@ -23,17 +23,20 @@ class ChatCompletionsAdapter:
             return obj
         if hasattr(obj, "model_dump"):
             return obj.model_dump()
+        if hasattr(obj, "__dict__"):
+            return vars(obj)
         return dict(obj)
 
     def _normalize_message(self, message: Any) -> NormalizedAssistantMessage:
         message_dict = self._to_dict(message)
         tool_calls = []
         for tc in message_dict.get("tool_calls") or []:
-            function = tc.get("function") or {}
+            tc_dict = self._to_dict(tc)
+            function = self._to_dict(tc_dict.get("function") or {})
             tool_calls.append(
                 NormalizedToolCall(
-                    id=tc.get("id", ""),
-                    type=tc.get("type", "function"),
+                    id=tc_dict.get("id", ""),
+                    type=tc_dict.get("type", "function"),
                     function=NormalizedToolFunction(
                         name=function.get("name", ""),
                         arguments=function.get("arguments", ""),
@@ -48,8 +51,9 @@ class ChatCompletionsAdapter:
 
     def _normalize_completion(self, response: Any) -> NormalizedCompletionResult:
         response_dict = self._to_dict(response)
-        choice = (response_dict.get("choices") or [{}])[0]
-        message = self._normalize_message(choice.get("message", {}))
+        raw_choices = response_dict.get("choices") or [{}]
+        choice = self._to_dict(raw_choices[0]) if raw_choices else {}
+        message = self._normalize_message(choice.get("message") or {})
         return NormalizedCompletionResult(
             message=message,
             finish_reason=choice.get("finish_reason"),
@@ -77,8 +81,9 @@ class ChatCompletionsAdapter:
     def _normalize_chunk_to_events(self, chunk: Any, index: int):
         events = []
         chunk_dict = self._to_dict(chunk)
-        choice = (chunk_dict.get("choices") or [{}])[0]
-        delta = choice.get("delta") or {}
+        raw_choices = chunk_dict.get("choices") or [{}]
+        choice = self._to_dict(raw_choices[0]) if raw_choices else {}
+        delta = self._to_dict(choice.get("delta") or {})
 
         content = delta.get("content")
         if content is not None:

